python/ray/serve/tests/test_new_handle_api.py

```diff
@@ -0,0 +1,288 @@
+            return await ref
+
+    serve.run(Deployment.bind(downstream.bind()))
+    handle = serve.get_app_handle("default")
+    assert isinstance(handle, DeploymentHandle)
+    assert handle.check_get_deployment_handle.remote().result() == "hello"
+
+    handle = serve.get_deployment_handle("downstream", app_name="default")
+    assert isinstance(handle, DeploymentHandle)
+    assert handle.remote().result() == "hello"
+
+
+def test_compose_deployments_in_app(serve_instance):
+    """Test composing deployment handle refs within a deployment."""
+
+    @serve.deployment
+    class Downstream:
+        def __init__(self, msg: str):
+            self._msg = msg
+
+        def __call__(self, inp: str):
+            return f"{self._msg}|{inp}"
+
+    @serve.deployment
+    class Deployment:
+        def __init__(self, handle1: RayServeHandle, handle2: RayServeHandle):
+            self._handle1 = handle1.options(use_new_handle_api=True)
+            self._handle2 = handle2.options(use_new_handle_api=True)
+
+        async def __call__(self):
+            result = await self._handle1.remote(self._handle2.remote("hi"))
+            return f"driver|{result}"
+
+    handle = serve.run(
+        Deployment.bind(
+            Downstream.options(name="downstream1").bind("downstream1"),
+            Downstream.options(name="downstream2").bind("downstream2"),
+        ),
+    ).options(use_new_handle_api=True)
+    assert handle.remote().result() == "driver|downstream1|downstream2|hi"
+
+
+def test_compose_apps(serve_instance):
+    """Test composing deployment handle refs outside of a deployment."""
+
+    @serve.deployment
+    class Deployment:
+        def __init__(self, msg: str):
+            self._msg = msg
+
+        def __call__(self, inp: str):
+            return f"{self._msg}|{inp}"
+
+    handle1 = serve.run(
+        Deployment.bind("app1"), name="app1", route_prefix="/app1"
+    ).options(use_new_handle_api=True)
+    handle2 = serve.run(
+        Deployment.bind("app2"), name="app2", route_prefix="/app2"
+    ).options(use_new_handle_api=True)
+
+    assert handle1.remote(handle2.remote("hi")).result() == "app1|app2|hi"
+
+
+def test_convert_to_object_ref(serve_instance):
+    """Test converting deployment handle refs to Ray object refs."""
+
+    @ray.remote
+    def identity_task(inp: Any):
+        return inp
+
+    @serve.deployment
+    def downstream():
+        return "hello"
+
+    @serve.deployment
+    class Deployment:
+        def __init__(self, handle: RayServeHandle):
+            self._handle = handle.options(use_new_handle_api=True)
+
+        async def __call__(self):
+            ref = self._handle.remote()
+            return await identity_task.remote(await ref._to_object_ref())
+
+    handle = serve.run(Deployment.bind(downstream.bind())).options(
+        use_new_handle_api=True
+    )
+
+    ref = handle.remote()
+    assert ray.get(identity_task.remote(ref._to_object_ref_sync())) == "hello"
```

Comment:
Why do we add an identity_task as part of this test?

Reply:
Just sanity checking that the obj ref can be passed to a Ray task!